Content.Server/GridPreloader/GridPreloaderSystem.cs

```diff
@@ -0,0 +1,114 @@
+            return;
+
+        preloader.Comp.PreloadGridsMapId = _mapManager.CreateMap();
+        _mapManager.AddUninitializedMap(preloader.Comp.PreloadGridsMapId.Value);
+        _mapManager.SetMapPaused(preloader.Comp.PreloadGridsMapId.Value, true);
+
+        var mapUid = _mapManager.GetMapEntityId(preloader.Comp.PreloadGridsMapId.Value);
+        var globalXOffset = 0f;
+        foreach (var proto in _prototype.EnumeratePrototypes<PreloadedGridPrototype>())
+        {
+            for (int i = 0; i < proto.Copies; i++)
+            {
+                var options = new MapLoadOptions
+                {
+                    LoadMap = false,
+                };
+
+                // i dont use TryLoad, because he doesn't return EntityUid
+                var gridUid = _map.LoadGrid(preloader.Comp.PreloadGridsMapId.Value, proto.Path.ToString(), options);
+                if (!TryComp<MapGridComponent>(gridUid, out var mapGrid))
+                    continue;
+
+                if (!TryComp<PhysicsComponent>(gridUid, out var physic))
+                    continue;
+
+                if (gridUid == null)
+                    continue;
+
+                //Position Calculating
+                globalXOffset += mapGrid.LocalAABB.Width / 2;
+
+                var coord = new Vector2(-physic.LocalCenter.X + globalXOffset, -physic.LocalCenter.Y);
+                _transform.SetCoordinates(gridUid.Value, new EntityCoordinates(mapUid, coord));
+
+                globalXOffset += (mapGrid.LocalAABB.Width / 2) + 1;
+
+                //Add to list
+                if (!preloader.Comp.PreloadedGrids.ContainsKey(proto.ID))
+                    preloader.Comp.PreloadedGrids[proto.ID] = new List<EntityUid>();
+                preloader.Comp.PreloadedGrids[proto.ID].Add(gridUid.Value);
+            }
+        }
+    }
+
+    public GridPreloaderComponent? GetPreloaderEntity()
+    {
+        return EntityQuery<GridPreloaderComponent>().FirstOrDefault();
+    }
+
+    /// <summary>
+    /// An attempt to get a certain preloaded shuttle. If there are no more such shuttles left, returns null
+    /// </summary>
+    public bool TryGetPreloadedGrid(ProtoId<PreloadedGridPrototype> proto, EntityCoordinates coord, out EntityUid? preloadedGrid, GridPreloaderComponent? preloader = null)
+    {
+        preloadedGrid = null;
+
+        if (preloader == null)
+        {
+            preloader = GetPreloaderEntity();
+            if (preloader == null)
+                return false;
+        }
+
+        if (preloader.PreloadedGrids.ContainsKey(proto) && preloader.PreloadedGrids[proto].Count > 0)
+        {
+            preloadedGrid = preloader.PreloadedGrids[proto][0];
+
+            _transform.SetCoordinates(preloadedGrid.Value, coord);
+
+            preloader.PreloadedGrids[proto].RemoveAt(0);
+            if (preloader.PreloadedGrids[proto].Count == 0)
+                preloader.PreloadedGrids.Remove(proto);
+
+            return true;
```

Comment:
This indexes into the dictionary a bunch instead of just re-using the value with trygetvalue.